source/utils/paragraphHelper.py

```diff
@@ -0,0 +1,284 @@
+			chunkStart = chunkEnd
+			startIndex = endIndex
+		else:
+			endIndex += 1
+	if chunkStart < paragraphLen:  # any leftovers?
+		yield paragraph[chunkStart:]
+
+
+def speakParagraph(ti: textInfos.TextInfo) -> None:
+	paragraph = ""
+	numLines = 0
+	tempTi = ti.copy()
+	while numLines < MAX_LINES:
+		tempTi.expand(textInfos.UNIT_LINE)
+		line = tempTi.text.strip()
+		if len(line) > 0:
+			paragraph += line + " "
+		if _isLastLineOfParagraph(tempTi.text):
+			break
+		if not tempTi.move(textInfos.UNIT_LINE, _Offset.NEXT_LINE):
+			break
+		numLines += 1
+
+	if len(paragraph.strip()) > 0:
+		gen = _splitParagraphIntoChunks(paragraph)
+		for chunk in gen:
+			speech.speakMessage(chunk)
+	else:
+		# Translators: This is spoken when a paragraph is considered blank.
+		speech.speakMessage(_("blank"))
+
+
+def moveToParagraph(nextParagraph: bool, speakNew: bool) -> Tuple[bool, bool]:
+	"""
+	Moves to the previous or next normal paragraph, delimited by a single line break.
+	@param nextParagraph: bool indicating desired direction of movement,
+	True for next paragraph, False for previous paragraph
+	@param speakNew: bool indicating if new paragraph should be spoken after navigating
+	@returns: A boolean 2-tuple of:
+	- passKey: if True, should send the gesture on
+	- moved: if True, position has changed
+	"""
+	ti = getTextInfoAtCaret()
+	if (ti is None) or (not _isAcceptableTextInfo(ti)):
+		return (True, False)
+	ti.expand(textInfos.UNIT_LINE)
+	ti.collapse()  # move to start of line
+	moveOffset: _Offset = _Offset.NEXT_LINE if nextParagraph else _Offset.PREVIOUS_LINE
+	moved = False
+	numLines = 0
+	tempTi = ti.copy()
+	tempTi.expand(textInfos.UNIT_LINE)
+	# if starting line is the last line of a paragraph, move back two paragraph markers
+	moveBackTwice = _isLastLineOfParagraph(tempTi.text)
+	while numLines < MAX_LINES:
+		tempTi = ti.copy()
+		tempTi.expand(textInfos.UNIT_LINE)
+		if _isLastLineOfParagraph(tempTi.text):
+			if not nextParagraph:
+				if not moveBackTwice:
+					while numLines < MAX_LINES:
+						if not ti.move(textInfos.UNIT_LINE, _Offset.PREVIOUS_LINE):
+							moved = True  # pin to beginning
+							break
+						tempTi = ti.copy()
+						tempTi.expand(textInfos.UNIT_LINE)
+						if _isLastLineOfParagraph(tempTi.text):
+							ti.move(textInfos.UNIT_LINE, _Offset.NEXT_LINE)
+							moved = True
+							break
+						numLines += 1
+
+					break  # break out of outer while loop
+				else:
+					moveBackTwice = False
+			else:  # moving to next paragraph
+				if ti.move(textInfos.UNIT_LINE, _Offset.NEXT_LINE):
+					moved = True
+				break
+		if not ti.move(textInfos.UNIT_LINE, moveOffset):
+			break
+		numLines += 1
+
+	if moved:
+		ti.updateCaret()
+		if isinstance(ti, UIATextInfo):
+			# Updating caret position in UIATextInfo does not scroll the display. Force it to scroll here.
+			ti._rangeObj.ScrollIntoView(False)
+		speakParagraph(ti)
+	else:
+		tones.beep(_ERROR_TONE.frequencyHz, _ERROR_TONE.durationMs)
```

Comment:
what are your thoughts on replace the beeps with a `ui.message` as Cyrille suggested earlier?

Reply:
I think messages instead of beeps would be find; I'll do it.

Reply:
Thanks